### DISCUSSION TITLE
How does duckdb generate the cli or .sql(...).show() previews? I would like to use it as function in a query.

### DISCUSSION THREAD

**Discussion by suiluj** [duckdb/duckdb#12322] [Q&A] [answered]
You get a text preview of a query result which shows longs cell values in a truncated format.

I would like to get these truncated values as extra viewing column.

So is it possible to use the method as general function which I can put infront of datatypes like `VARCHAR` and `BLOB`? So I would be able to always process a query as viewable version even for long values.

**Comment by Tishj** [comment]
```c++
string DuckDBPyRelation::ToStringInternal(const BoxRendererConfig &config, bool invalidate_cache) {
	AssertRelation();
	if (rendered_result.empty() || invalidate_cache) {
		BoxRenderer renderer;
		auto limit = Limit(config.limit, 0);
		auto res = limit->ExecuteInternal();

		auto context = rel->context.GetContext();
		rendered_result = res->ToBox(*context, config);
	}
	return rendered_result;
}

string DuckDBPyRelation::ToString() {
	BoxRendererConfig config;
	config.limit = 10000;
	return ToStringInternal(config);
}
```

Assuming you are referring to the python `con.sql(...).show()`

**Reply by Tishj** [reply]
It essentially does a LIMIT clause, and then it uses the BoxRenderer to render the result, the CLI also uses the same BoxRenderer code

**Reply by suiluj** [reply]
okay but I do not understand how to call this as a function in a query

I would like to do something like this:

- I want to generate simple table previews for every kind of table data manually. So I do not want to use the python or cli `.show()` BoxRenderer but use this internal `ToString()` it as part of a query myself
- lets say I have a table with twoch columns
- column 1 of type `VARCHAR` that can contain very long texts
- column 2 of type `BLOB` that contains bytes that could sometimes be converted to strings so it would be interesting to have a preview text in that case

I would like to use a query similar to this:

```sql
select
  preview_function(column_1) as column_1_preview, 
  preview_fucntion(column_2) as column_2_preview
from example_table
```

It would be great to have this generic `preview_function` that could work with every type of column similar as the pyhton `.con.sql(...).show()` already does. But I do not want to use some kind of markdown renderer but plain sql.

Is there a simple way to do it in a generic way?

I already thought about [try_cast](https://duckdb.org/docs/sql/expressions/cast.html#try_cast) to a string and then `left(content, 100) as content_preview` but this would involve more manual adjustments than a simple generic preview method.

**Comment by suiluj** [comment, marked as answer]
Ah I just got an idea for a great macro! :)

```sql
create or replace macro preview(value,preview_length) as left(try_cast( value as varchar), preview_length);
```

Then I can create a query like this:

```sql
select
  preview(column_1, 5) as column_1_preview,
  preview(column_2, 5) as column_2_preview
from example_table
```

I think this solutions works for my use case.

**Reply by suiluj** [reply]
some improvements for the preview macro:

```sql
create or replace macro preview(value, length := 20, prefix := '', suffix := '') as concat(prefix, left(try_cast(value as varchar), length), suffix);
```

```sql
select  
  preview(column_1) as column_1_preview,
  preview(column_1, length := 3, prefix := 'start: ', suffix := ' (end)') as column_1_custom_preview,
from example_table
```
